### DISCUSSION TITLE
Do you perform conditional redirection in a layout, or in every single loader (and action)?

### DISCUSSION THREAD

**Discussion by n1ghtmare** [remix-run/remix#4460] [Q&A] [answered]
Hi, everyone, I'm trying to understand how to structure my Remix application. My current approach has some weird race conditions (due to the fact that loaders run in parallel), and I'm getting a bit confused.

This is my scenario:

I have a bunch of pages with a shared layout that don't require authentication. For example:

 - `__main.tsx` (used as a layout)
 - `__main/signin.tsx` (signin with "magic link" through email)
 - `__main/verify-request.tsx` (signin will redirect here after submission)
 - `__main/email-callback.tsx` (used for when the user has clicked the "magic link")
 - `... more pages ...`

Then a few pages that do require authentication, such as:

 - `__internal.tsx` (used as a layout)
 - `__internal/dashboard.tsx`
 - `__internal/dashboard/widget.tsx`
 - `__internal/dashboard/widget/create.tsx`
 - `... more pages ...`

I currently perform the following checks:

 - `__main.tsx` -> if the user has a session, redirect to /dashboard.
 - `__main.tsx` -> if the user has requested a "magic link" (I check a session value) redirect to `/verify-request` (which displays a message, that the user has to check their email and use the link).
 - `__internal.tsx` -> if the user doesn't have a session, redirect to `/signin`.

Now here is the problem, once the user clicks the magic link (from his inbox), and navigates to `/email-callback`, the loaders for both `__main.tsx` (layout) and `__main/email-callback.tsx` are executed in parallel. So from `/email-callback`, I validate the link, if valid, I redirect the user to `/dashboard`, however before this happens, the loader for `__main.tsx` redirects the user to `/verify-request`.

I understand why this happens (I think). Should I perform these checks in every single loader (and action?)? Even deeply nested routes, such as `__internal/dashboard/widget/create.tsx`? Or keep this strategy, but structure my layouts in a different way? What's the idiomatic way to handle this? I really appreciate some help here.

**Comment by lili21** [comment]
how about do it in `CatchBoundary`

**Reply by n1ghtmare** [reply]
Thank you for replying, I appreciate it!

Do you mean, check if the user has a session, if he doesn't throw an error, and redirect in the `CatchBoundary` accordingly? If this is what you mean, my question remains, do I perform the checks for user session in all the loaders or just the parent "layout" route loader?

**Reply by lili21** [reply]
in this way you don't need check it for all loaders, just the parent would be enough.

**Reply by sergiodxa** [reply]
CatchBoundary is an UI, you can't do proper redirect there, you may need to use the useNavigate hook and navigate to another route, but the user would have already received a non-redirect response

**Comment by machour** [comment, marked as answer]
Given the fact that loaders are called in parallel, you should check for authentication in **every** route loader.

**Reply by n1ghtmare** [reply]
Thank you, I thought as much, but couldn't figure out a scenario where this could be a problem. Now that I think about it, it makes sense tbh, you don't want to execute code that is dependent on the parent loader performing a check (things might get executed in a different order than you're betting on)...

I appreciate the help, just wanted to make sure if I'm doing things the "idiomatic way".

**Reply by machour** [reply]
I could call your child loader url using the remix ?_data parameter and defeat your parent guard. Or it could be defeated by just calling the URL if the child loader gets executed first. 

Until remix have something like middleware, guard everything thoroughly 👮‍♂️

**Reply by n1ghtmare** [reply]
Oh, I didn't think about these scenarios. Thank you, this actually really helped me. It's a pain to refactor everything and perform the same checks all over the place, but, oh well. Some helper functions here and there should do the trick.

**Reply by machour** [reply]
Make your helper throw a Response if unauthorized, you'll have a one-liner to plug in all your loaders. 
Your CatchBoundary will handle the thrown response!

**Reply by sergiodxa** [reply]
Another thing that could happen and most likely will happen is that when performing a client-side navigation Remix will only call the changing route loaders, this means if you check for auth in the root route loader that one is never going to be called again until an action triggers a revalidation so your other loaders will be unprotected in that case.

Another problem is that your nested loaders may need to know about the authenticated user to get data the user is authorized to access, so you still need to get the user data there.

**Reply by n1ghtmare** [reply]
@sergiodxa Thank you for taking the time to answer, I learned a lot. I appreciate it!
